subsys/net/ip/net_if.c

```diff
@@ -226,7 +226,11 @@ static inline void init_iface(struct net_if *iface)
 
 	NET_DBG("On iface %p", iface);
 
-	api->init(iface);
+	if (api && api->init) {
+		api->init(iface);
+	} else {
+		NET_ERR("init fun NULL on iface %p", iface);
```

Comment:
Perhaps s/fun/func/ as that is more clearer.
I am a bit surprised that there is no other errors later then the network interface code is running for that interface. How much did you tested this?

Reply:
I just tried this with qemu_x86 (in `samples/net/sockets/echo_server`, build command is `cmake .. -DBOARD=qemu_x86 -DOVERLAY_CONFIG=overlay-e1000.conf` with this modification
```
diff --git a/subsys/net/ip/net_if.c b/subsys/net/ip/net_if.c
index 3d3fc52b4a..266b18a5c2 100644
--- a/subsys/net/ip/net_if.c
+++ b/subsys/net/ip/net_if.c
@@ -217,6 +217,9 @@ static inline void init_iface(struct net_if *iface)
 {
        const struct net_if_api *api = net_if_get_device(iface)->driver_api;
 
+       net_if_get_device(iface)->driver_api = NULL;
+       return;
+
        NET_ASSERT(api && api->init);
 
        NET_DBG("On iface %p", iface);
```

This crash is observed
```
SeaBIOS (version rel-1.12.0-0-ga698c8995f-prebuilt.qemu.org)

iPXE (http://ipxe.org) 00:03.0 C980 PCI2.10 PnP PMM+007916B0+006F16B0 C980

Booting from ROM..***** CPU Page Fault (error code 0x00000000)
Supervisor thread read address 0x00000008
PDE: 0x025 Present, Read-only, User, Execute Enabled
PTE: 0x00 Non-present, Read-only, Supervisor, Execute Enabled
Current thread ID = 0x00401fa0
eax: 0x0040f9e0, ebx: 0x0040f920, ecx: 0x00401000, edx: 0x0040f624
esi: 0x00000000, edi: 0x00002781, ebp: 0x0040b444, esp: 0x0040b43c
eflags: 0x00000202 cs: 0x0008
call trace:
eip: 0x0000ba8d
     0x0000f391 (0x40f920)
     0x0000f5b4 (0x40f920)
     0x0000d1c6 (0x40b484)
     0x00019ccb (0x40f630)
     0x00019d51 (0x2)
     0x00002790 (0x0)
Fatal fault in essential thread! Spinning...
Terminate emulator due to fatal kernel error
make[3]: *** [zephyr/CMakeFiles/run.dir/build.make:60: zephyr/CMakeFiles/run] Error 1
make[2]: *** [CMakeFiles/Makefile2:401: zephyr/CMakeFiles/run.dir/all] Error 2
make[1]: *** [CMakeFiles/Makefile2:408: zephyr/CMakeFiles/run.dir/rule] Error 2
make: *** [Makefile:222: run] Error 2
```

It looks like we need to add some more checks, than this simple one, for this to work reliably.

Reply:
The crash I was seeing was caused by the driver_api pointer being used in `subsys/net/l2/ethernet/ethernet.c` so we need to fix all those checks in various files.

```
$ ack --cc driver_api subsys/net/

subsys/net/lib/sockets/sockets_can.c
355:		api = dev->driver_api;
393:		api = dev->driver_api;

subsys/net/lib/openthread/platform/radio.c
73:	radio_api = (struct ieee802154_radio_api *)radio_dev->driver_api;

subsys/net/l2/dummy/dummy.c
32:	const struct dummy_api *api = net_if_get_device(iface)->driver_api;

subsys/net/l2/ethernet/ethernet_mgmt.c
19:	const struct ethernet_api *api = dev->driver_api;
30:	const struct ethernet_api *api = dev->driver_api;
172:	const struct ethernet_api *api = dev->driver_api;

subsys/net/l2/ethernet/ethernet_stats.c
33:		eth = net_if_get_device(iface)->driver_api;

subsys/net/l2/ethernet/ethernet.c
513:	const struct ethernet_api *api = net_if_get_device(iface)->driver_api;
610:		net_if_get_device(iface)->driver_api;
772:		net_if_get_device(iface)->driver_api;
834:		net_if_get_device(iface)->driver_api;
933:	const struct ethernet_api *api = dev->driver_api;

subsys/net/l2/ethernet/eth_stats.h
20:		net_if_get_device(iface)->driver_api;
39:		net_if_get_device(iface)->driver_api;
57:		net_if_get_device(iface)->driver_api;
75:		net_if_get_device(iface)->driver_api;
93:		net_if_get_device(iface)->driver_api;
111:		net_if_get_device(iface)->driver_api;
129:		net_if_get_device(iface)->driver_api;
147:		net_if_get_device(iface)->driver_api;
167:		net_if_get_device(iface)->driver_api);
185:		net_if_get_device(iface)->driver_api);

subsys/net/l2/canbus/canbus.c
33:	const struct canbus_api *api = net_if_get_device(iface)->driver_api;

subsys/net/l2/wifi/wifi_mgmt.c
23:		(struct net_wifi_mgmt_offload *) dev->driver_api;
78:		(struct net_wifi_mgmt_offload *) dev->driver_api;
95:		(struct net_wifi_mgmt_offload *) dev->driver_api;
135:		(struct net_wifi_mgmt_offload *) dev->driver_api;
151:		(struct net_wifi_mgmt_offload *) dev->driver_api;

subsys/net/l2/ieee802154/ieee802154_utils.h
21:		net_if_get_device(iface)->driver_api;
29:		net_if_get_device(iface)->driver_api;
37:		net_if_get_device(iface)->driver_api;
45:		net_if_get_device(iface)->driver_api;
54:		net_if_get_device(iface)->driver_api;
62:		net_if_get_device(iface)->driver_api;
70:		net_if_get_device(iface)->driver_api;
79:		net_if_get_device(iface)->driver_api;
99:		net_if_get_device(iface)->driver_api;
119:		net_if_get_device(iface)->driver_api;
139:		net_if_get_device(iface)->driver_api;
159:		net_if_get_device(iface)->driver_api;
179:		net_if_get_device(iface)->driver_api;
199:		net_if_get_device(iface)->driver_api;
224:		net_if_get_device(iface)->driver_api;
```

So we would need to add checks for null driver_api in each of the lines above. Probably returning `-EIO` would be enough in that case. @vonhust could you enhance this PR by adding a commit that adds suitable checks?

Reply:
Wow, so many places. To be honest, I am not good at zephyr's network stack. So difficult to provide a high quality solution.

BTW, to avoid so many changes in network, maybe we can consider do some change in `z_sys_device_do_config_level`  to avoid info->driver_api = NULL

Reply:
> BTW, to avoid so many changes in network, maybe we can consider do some change in z_sys_device_do_config_level to avoid info->driver_api = NULL

Would have been nice in the first place not to get that change. Ideally struct device should be const (as well as its attributes) and placed into ROM.
This driver_api set to NULL is a quick "fix" to the issue of driver not initializing properly.

Reply:
> Wow, so many places. To be honest, I am not good at zephyr's network stack. So difficult to provide a high quality solution.

It should not be that difficult, just some checks in the beginning of the funcs. I can also provide a PR if you like, just let me know.

Reply:
@jukkar , it's better that you provide a PR as you are an expert in zephyr's network stack

Reply:
Ok, I can do that.